core/driver/sources/src/main/scala/org/virtuslab/ideprobe/ide/intellij/IntelliJProvider.scala

```diff
@@ -0,0 +1,152 @@
+      PluginArchive(plugin, file.toArchive)
+    })
+
+    val distinctPlugins = archives.reverse.distinctBy(_.rootEntries).reverse
+
+    parallel(distinctPlugins).forEach { pluginArchive =>
+      pluginArchive.archive.extractTo(targetDir)
+      println(s"Installed ${pluginArchive.plugin}")
+    }
+  }
+
+  private def withParallel[A, B](s: Seq[A])(f: JStream[A] => JStream[B]): Seq[B] = {
+    f(parallel(s)).collect(Collectors.toList[B]).asScala.toList
+  }
+
+  private def parallel[A](s: Seq[A]) = {
+    s.asJava.parallelStream
+  }
+}
+
+final class ExistingIntelliJ(
+    dependencies: DependencyProvider,
+    path: Path,
+    override val plugins: Seq[Plugin],
+    val paths: IdeProbePaths,
+    val config: DriverConfig
+) extends IntelliJProvider {
+  override val version = IntelliJVersionResolver.version(path)
+  override def withConfig(config: DriverConfig): ExistingIntelliJ =
+    new ExistingIntelliJ(dependencies, path, plugins, paths, config)
+
+  override def withPaths(paths: IdeProbePaths): ExistingIntelliJ =
+    new ExistingIntelliJ(dependencies, path, plugins, paths, config)
+
+  override def withPlugins(plugins: Plugin*): ExistingIntelliJ =
+    new ExistingIntelliJ(dependencies, path, this.plugins ++ plugins, paths, config)
+
+  override def setup(): InstalledIntelliJ = {
+    val pluginsDir = path.resolve("plugins")
+    val backupDir = Files.createTempDirectory(path, "plugins")
+    pluginsDir.copyDir(backupDir)
+    installPlugins(dependencies, plugins, path)
+
+    new LocalIntelliJ(path, paths, config, backupDir)
+  }
+}
+
+final class IntelliJFactory(
+    dependencies: DependencyProvider,
+    override val plugins: Seq[Plugin],
+    override val version: IntelliJVersion,
+    val paths: IdeProbePaths,
+    val config: DriverConfig
+) extends IntelliJProvider {
+  override def withConfig(config: DriverConfig): IntelliJFactory =
+    new IntelliJFactory(dependencies, plugins, version, paths, config)
+
+  override def withPaths(paths: IdeProbePaths): IntelliJFactory =
+    new IntelliJFactory(dependencies, plugins, version, paths, config)
+
+  override def withPlugins(plugins: Plugin*): IntelliJProvider =
+    new IntelliJFactory(dependencies, plugins = this.plugins ++ plugins, version, paths, config)
+
+  override def setup(): InstalledIntelliJ = {
+    val root = createInstanceDirectory(version)
+
+    installIntelliJ(version, root)
+    installPlugins(dependencies, plugins, root)
+
+    new DownloadedIntelliJ(root, paths, config)
+  }
+
+  private def createInstanceDirectory(version: IntelliJVersion): Path = {
+    val path = paths.instances.createTempDirectory(s"intellij-instance-${version.build}-")
+
+    Files.createDirectories(path)
+  }
+
+  private def installIntelliJ(version: IntelliJVersion, root: Path): Unit = {
+    println(s"Installing $version")
+    val file = dependencies.fetch(version)
+    file.toArchive.extractTo(root)
+    root.resolve("bin/linux/fsnotifier64").makeExecutable()
```

Comment:
followup/nice to have - it would be good to traverse bin dir and make everything executable. We have some mess in logs because we don't do this.

Reply:
Done.